articles/ros_node_idl.md

```diff
@@ -0,0 +1,182 @@
+
+These examples are only a subset of use-cases made possible by such an interface.
+It's clear that this is useful well beyond security.
+
+## Package interface
+
+How do upstream packages specify their interface requirements?
+Through a high-level description of all the actions, parameters, services and topics, provided or required, by each node within the package.
+
+The package interface is defined in a separate [XML][xml_wiki] file and exported from the `package.xml` using [REP 149's export mechanism][rep149_export], thereby avoiding pollution of the package manifest.
+The interface may cover only a subset of nodes in a package, as long as the nodes that _are_ covered are done so completely.
+
+Here is an example IDL for a package containing two nodes:
+
+{% include_relative ros2_node_idl/interface_declaration.xml %}
+
+Once an IDL file is written, it is exported from the package manifest:
+
+{% include_relative ros2_node_idl/package.xml %}
+
+### Schema for `package.xml`'s export tag
+
+#### `interface`
+
+This is how the package exports its defined IDL for other tools to consume.
+
+Attributes:
+
+**path**:  Path to XML file containing IDL
+
+Note that the introduction of the `interface` tag within the `export` tag of the package manifest raises a small difficulty with regards to the [REP 149][rep149_export].
+The REP specifies:
+
+> To avoid potential collisions, an export tag should have the same name as the package which is meant to process it. The content of that tag is up to the package to define and use.
+
+Considering the high level abstraction of the IDL, the `interface` tag is not meant for a specific package but rather declares intrinsic properties for anyone to process it.
+However, the keyword is likely solidly descriptive enough to either be accepted as falling under [REP 149][rep149_export] or else to motivate an amendment of the REP.
+
+### IDL Schema
+
+#### `interface`
+
+This tag contains the package interface, made up of a collection of node interfaces.
+Root tag of the IDL file. There must be only one <interface> tag per IDL file.
+
+Attributes:
+- **version**: version of schema in use, allowing for future revisions.
+
+#### `node`
+
+Encapsulate a sequence of ROS entity interfaces.
+It is specific to a node as determined by its associated attributes.
+
+Attributes:
+- **name**: The base name of the node.
+
+#### `action`
+
+Define the interface for a given action.
+
+Attributes:
+- **name**: The name of the action.
+- **type**: The type of the action.
+Valid values are any ROS action types.
+- **server**: Whether or not the node provides an action server for the action.
+Valid values are "true" or "false". Defaults to "false".
+- **client**: Whether or not the node provides a client for the action.
+Valid values are "true" or "false". Defaults to "false".
+
+#### `parameter`
+
+Define the interface for a given parameter.
+
+Attributes:
```

Comment:
ROS 2 introduced [read only parameters](https://github.com/ros2/rclcpp/pull/495). Should the parameters attributes includes 'ro/rw' ?

Reply:
IMO, it's okay to not include the read only status. We can already query this constraint on parameters (and other constraints) at runtime. Unless there's a strong case for static analysis, I feel like describing it in the node IDL just adds another point where information can go out-of-sync.

Reply:
It might be good if the parameter description could be separated. I could see it being read independently by a node which doesn't have a "Node IDL" but wants to read the declared parameters from a file. That would imply that the parameter would need several additional arguments though (all which are possible to pass using the parameter declaration API: ranges, read-only, ...).

Reply:
@jacobperron Not that one of the point of this proposal is to access the API info in a 'pre-runtime' fashion - e.g. at launch-parsing time. So querying for ro/rw and other constraints at runtime is kinda too late already.

@dirk-thomas I agree that it might be interesting, but it does feel a little out of scope here. Do you mean something a la ROS 1 [rosparam_handler](https://github.com/cbandera/rosparam_handler/blob/develop/doc/HowToWriteYourFirstParamsFile.md)?